[UnifiedPDF] Stop using deprecated PDFAnnotation subclasses

aprotyas opened the pull request:
#### 375adeb78a3cbc278b08b358ad950805be8692d9
<pre>
[UnifiedPDF] Stop using deprecated PDFAnnotation subclasses
<a href="https://bugs.webkit.org/show_bug.cgi?id=279807">https://bugs.webkit.org/show_bug.cgi?id=279807</a>
<a href="https://rdar.apple.com/136120943">rdar://136120943</a>

Reviewed by Sammy Gill and Aditya Keerthi.

To interact with the various widget subtypes (button, choice, text), we
were directly casting to (and calling into) sublasses of PDFAnnotation,
namely PDFAnnotation[Button|Choice|Text]Widget. These subclasses are
deprecated and (more importantly) unavailable in iOS. The latter issue
stands in the way of making PDFPluginAnnotation a cross-platform
concept, which we need to do in anticipation of password-protected PDF
support for the iOS family.

As such, this patch moves us away from the deprecated subclasses to
instead query widget types through the key-based interface defined by
PDFAnnotation[Key|[Widget]Subtype]. Moreover, it uses appropriate
replacement methods defined in the PDFAnnotationUtilities protocol
category.

* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:

Stop soft linking the deprecated PDFAnnotation subclasses.
Start soft linking the keys of interest.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.h: Renamed from Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTextWidgetDetails.h.
* Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.mm: Added.
(WebKit::PDFAnnotationTypeHelpers::widgetType):
(WebKit::PDFAnnotationTypeHelpers::annotationIsWidget):
(WebKit::PDFAnnotationTypeHelpers::annotationIsWidgetOfType):

Make PDF annotation (widget) type querying ergonomic through a set of
helper functions.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::setActiveAnnotation):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::create):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.h:
(WebKit::PDFPluginChoiceAnnotation::choiceAnnotation): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.mm:
(WebKit::PDFPluginChoiceAnnotation::create):
(WebKit::PDFPluginChoiceAnnotation::updateGeometry):
(WebKit::PDFPluginChoiceAnnotation::commit):
(WebKit::PDFPluginChoiceAnnotation::createAnnotationElement):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginTextAnnotation.h:
(WebKit::PDFPluginTextAnnotation::textAnnotation const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginTextAnnotation.mm:
(WebKit::PDFPluginTextAnnotation::create):
(WebKit::PDFPluginTextAnnotation::createAnnotationElement):
(WebKit::PDFPluginTextAnnotation::updateGeometry):
(WebKit::PDFPluginTextAnnotation::commit):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pageIndexWithHoveredAnnotation const):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::repaintRequirementsForAnnotation):
(WebKit::findFirstTextAnnotationStartingAtIndex):
(WebKit::UnifiedPDFPlugin::setActiveAnnotation):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:

Unified sources fixes.

Canonical link: <a href="https://commits.webkit.org/283801@main">https://commits.webkit.org/283801@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/86bcc5eedd504fb819a26bdbcdd290129cdc5bcc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67415 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46794 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/20047 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18552 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54592 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18343 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/71463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12432 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70482 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42969 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58313 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/71463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39642 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16910 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16055 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73162 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11374 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/15377 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/61483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11409 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58379 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/61542 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9301 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2915 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10240 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42599 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43676 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44862 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43417 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.mm:
```diff
@@ -0,0 +1,80 @@
+
+bool annotationIsWidget(PDFAnnotation *annotation)
+{
+    return [[annotation valueForAnnotationKey:get_PDFKit_PDFAnnotationKeySubtype()] isEqualToString:get_PDFKit_PDFAnnotationSubtypeWidget()] == YES;
```
I think we use `static_cast<bool>` more commonly?

aprotyas replied:
Okay!

aprotyas replied:
Fixed in [116c479](https://github.com/WebKit/WebKit/commit/116c47933b9948f1c2e49242b9a652d27f3dee49)

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.mm:
```diff
@@ -52,22 +55,20 @@
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    PDFAnnotationChoiceWidget *choiceAnnotation = this->choiceAnnotation();
-ALLOW_DEPRECATED_DECLARATIONS_END
+    PDFAnnotation *choiceAnnotation = annotation();
```
`RetainPtr`

aprotyas replied:
Fixed in [116c479](https://github.com/WebKit/WebKit/commit/116c47933b9948f1c2e49242b9a652d27f3dee49)

pxlcoder commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginTextAnnotation.mm:
```diff
@@ -81,9 +83,7 @@ static const String cssAlignmentValueForNSTextAlignment(NSTextAlignment alignmen
-ALLOW_DEPRECATED_DECLARATIONS_BEGIN
-    PDFAnnotationTextWidget *textAnnotation = this->textAnnotation();
-ALLOW_DEPRECATED_DECLARATIONS_END
+    PDFAnnotation *textAnnotation = annotation();
```
`RetainPtr`

aprotyas replied:
Fixed in [116c479](https://github.com/WebKit/WebKit/commit/116c47933b9948f1c2e49242b9a652d27f3dee49)

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.mm:
```diff
@@ -0,0 +1,80 @@
+    }).value_or(false);
+}
+
+bool annotationIsWidgetOfType(PDFAnnotation *annotation, std::initializer_list<WidgetType>&& types)
```
Nit: I feel like the naming of this helper could somehow indicate that we are checking to see if the annotation is one of the passed in types. I think this would help improve the readability/clarity of what we are doing at the call sites. WDYT?

aprotyas replied:
Yeah, this can be confusing... maybe I'll just get rid of this and have callers break up calls like this:

```cpp
annotationIsWidgetOfType(annotation, { WidgetType::Text, WidgetType::Button });
```

... with calls like this:

```cpp
annotationIsWidgetOfType(annotation, WidgetType::Text) || annotationIsWidgetOfType(annotation, WidgetType::Button);
```

aprotyas committed:
[UnifiedPDF] Stop using deprecated PDFAnnotation subclasses
https://bugs.webkit.org/show_bug.cgi?id=279807
rdar://136120943

Reviewed by Sammy Gill and Aditya Keerthi.

To interact with the various widget subtypes (button, choice, text), we
were directly casting to (and calling into) sublasses of PDFAnnotation,
namely PDFAnnotation[Button|Choice|Text]Widget. These subclasses are
deprecated and (more importantly) unavailable in iOS. The latter issue
stands in the way of making PDFPluginAnnotation a cross-platform
concept, which we need to do in anticipation of password-protected PDF
support for the iOS family.

As such, this patch moves us away from the deprecated subclasses to
instead query widget types through the key-based interface defined by
PDFAnnotation[Key|[Widget]Subtype]. Moreover, it uses appropriate
replacement methods defined in the PDFAnnotationUtilities protocol
category.

* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:

Stop soft linking the deprecated PDFAnnotation subclasses.
Start soft linking the keys of interest.

* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.h: Renamed from Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTextWidgetDetails.h.
* Source/WebKit/WebProcess/Plugins/PDF/PDFAnnotationTypeHelpers.mm: Added.
(WebKit::PDFAnnotationTypeHelpers::widgetType):
(WebKit::PDFAnnotationTypeHelpers::annotationIsWidget):
(WebKit::PDFAnnotationTypeHelpers::annotationIsWidgetOfType):

Make PDF annotation (widget) type querying ergonomic through a set of
helper functions.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPlugin.mm:
(WebKit::PDFPlugin::setActiveAnnotation):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginAnnotation.mm:
(WebKit::PDFPluginAnnotation::create):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.h:
(WebKit::PDFPluginChoiceAnnotation::choiceAnnotation): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginChoiceAnnotation.mm:
(WebKit::PDFPluginChoiceAnnotation::create):
(WebKit::PDFPluginChoiceAnnotation::updateGeometry):
(WebKit::PDFPluginChoiceAnnotation::commit):
(WebKit::PDFPluginChoiceAnnotation::createAnnotationElement):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginTextAnnotation.h:
(WebKit::PDFPluginTextAnnotation::textAnnotation const): Deleted.
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginTextAnnotation.mm:
(WebKit::PDFPluginTextAnnotation::create):
(WebKit::PDFPluginTextAnnotation::createAnnotationElement):
(WebKit::PDFPluginTextAnnotation::updateGeometry):
(WebKit::PDFPluginTextAnnotation::commit):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFPresentationController.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFScrollingPresentationController.mm:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::pageIndexWithHoveredAnnotation const):
(WebKit::UnifiedPDFPlugin::pdfElementTypesForPluginPoint const):
(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::repaintRequirementsForAnnotation):
(WebKit::findFirstTextAnnotationStartingAtIndex):
(WebKit::UnifiedPDFPlugin::setActiveAnnotation):
(WebKit::AnnotationTrackingState::startAnnotationTracking):
(WebKit::AnnotationTrackingState::finishAnnotationTracking):

* Source/WebKit/WebProcess/WebPage/Cocoa/TextAnimationController.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebCookieJarCocoa.mm:
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:

Unified sources fixes.

Canonical link: https://commits.webkit.org/283801@main
